File: apps/app/modules/expo-pose-landmarker/ios/ExpoPoseLandmarkerModule.swift
import ExpoModulesCore
import MediaPipeTasksVision
import UIKit

// MediaPipe Pose Landmarker → COCO-17 landmark name mapping. Mirrors
// MEDIAPIPE_TO_COCO17 in landmarkMapping.ts. Indices unmapped here are dropped
// before being sent across the bridge (we only need the 17 keypoints the
// athmira overlay math expects).
private let mediaPipeIndexToCoco17: [Int: String] = [
  0: "nose",
  2: "left_eye",
  5: "right_eye",
  7: "left_ear",
  8: "right_ear",
  11: "left_shoulder",
  12: "right_shoulder",
  13: "left_elbow",
  14: "right_elbow",
  15: "left_wrist",
  16: "right_wrist",
  23: "left_hip",
  24: "right_hip",
  25: "left_knee",
  26: "right_knee",
  27: "left_ankle",
  28: "right_ankle"
]

public class ExpoPoseLandmarkerModule: Module {
  private var landmarker: PoseLandmarker?
  private var activeDelegate: Delegate = .GPU
  private let modelAssetName = "pose_landmarker_lite"

  public func definition() -> ModuleDefinition {
    Name("ExpoPoseLandmarkerModule")

    OnCreate {
      do {
        try self.initializeLandmarker()
      } catch {
        NSLog("[ExpoPoseLandmarker] init failed: \(error.localizedDescription)")
      }
    }

    AsyncFunction("detectPoseFromUri") { (uri: String, promise: Promise) in
      guard let landmarker = self.landmarker else {
        promise.reject("E_NOT_READY", "PoseLandmarker is not initialized")
        return
      }

      guard let image = self.loadImage(fromUri: uri) else {
        promise.reject("E_IMAGE_LOAD", "Could not load image at \(uri)")
        return
      }

      guard let mpImage = try? MPImage(uiImage: image) else {
        promise.reject("E_MP_IMAGE", "Could not create MPImage from UIImage")
        return
      }

      let startedAt = Date()

      do {
        let result = try landmarker.detect(image: mpImage)
        let inferenceMs = Date().timeIntervalSince(startedAt) * 1000.0

        guard let firstPose = result.landmarks.first else {
          promise.resolve(nil)
          return
        }

        var landmarks: [[String: Any]] = []
        for (index, landmark) in firstPose.enumerated() {
          guard let name = mediaPipeIndexToCoco17[index] else { continue }
          // MediaPipe presence/visibility are NSNumber? in the Swift bridge.
          // Prefer presence (model confidence the landmark exists);
          // fall back to visibility if presence is nil.
          let confidence = landmark.presence?.doubleValue
            ?? landmark.visibility?.doubleValue
            ?? 0.0
          landmarks.append([
            "name": name,
            "x": Double(landmark.x),
            "y": Double(landmark.y),
            "confidence": confidence
          ])
        }

        promise.resolve([
          "landmarks": landmarks,
          "imageWidth": Double(image.size.width * image.scale),
          "imageHeight": Double(image.size.height * image.scale),
          "inferenceMs": inferenceMs,
          "delegate": self.activeDelegate == .GPU ? "gpu" : "cpu"
        ])
      } catch {
        promise.reject("E_DETECT", error.localizedDescription)
      }
    }
    .runOnQueue(.global(qos: .userInitiated))

    AsyncFunction("warmUp") { (promise: Promise) in
      guard let landmarker = self.landmarker else {
        promise.resolve(nil)
        return
      }
      // Build a 192x192 black UIImage and run one detection to prime the
      // Metal pipeline. First inference cost is ~200-400 ms; subsequent
      // calls are ~15-25 ms.
      let warmUpImage = self.makeBlackImage(side: 192)
      if let mpImage = try? MPImage(uiImage: warmUpImage) {
        _ = try? landmarker.detect(image: mpImage)
      }
      promise.resolve(nil)
    }
    .runOnQueue(.global(qos: .userInitiated))

    Function("getActiveDelegate") { () -> String in
      return self.activeDelegate == .GPU ? "gpu" : "cpu"
    }

    // M2: live-stream view. Owns the AVCaptureSession, runs MediaPipe in
    // `.liveStream` mode, emits poses directly via the `onPose` event.
    // See PoseLandmarkerView.swift for the implementation.
    View(PoseLandmarkerView.self) {
      Events("onPose", "onReady", "onMountError")

      Prop("facing") { (view: PoseLandmarkerView, value: String) in
        view.facing = value
      }
      Prop("mirror") { (view: PoseLandmarkerView, value: Bool) in
        view.mirror = value
      }
      Prop("enabled") { (view: PoseLandmarkerView, value: Bool) in
        view.enabled = value
      }

      AsyncFunction("takePicture") { (view: PoseLandmarkerView, promise: Promise) in
        view.takePicture(promise: promise)
      }
    }
  }

  // MARK: - Private helpers

  private func initializeLandmarker() throws {
    let options = PoseLandmarkerOptions()
    guard
      let modelPath = Bundle.main.path(forResource: modelAssetName, ofType: "task")
        ?? Bundle(for: type(of: self)).path(forResource: modelAssetName, ofType: "task")
    else {
      throw NSError(
        domain: "ExpoPoseLandmarker",
        code: 1,
        userInfo: [NSLocalizedDescriptionKey: "Bundled \(modelAssetName).task not found"]
      )
    }
    options.baseOptions.modelAssetPath = modelPath
    options.baseOptions.delegate = .GPU
    options.runningMode = .image
    options.numPoses = 1
    options.minPoseDetectionConfidence = 0.5
    options.minPosePresenceConfidence = 0.5
    options.minTrackingConfidence = 0.5

    do {
      self.landmarker = try PoseLandmarker(options: options)
      self.activeDelegate = .GPU
    } catch {
      // Some older devices reject the GPU delegate. Fall back to CPU.
      NSLog("[ExpoPoseLandmarker] GPU delegate failed, falling back to CPU: \(error.localizedDescription)")
      options.baseOptions.delegate = .CPU
      self.landmarker = try PoseLandmarker(options: options)
      self.activeDelegate = .CPU
    }
    NSLog("[ExpoPoseLandmarker] initialized with delegate: \(self.activeDelegate == .GPU ? "GPU" : "CPU")")
  }

  private func loadImage(fromUri uri: String) -> UIImage? {
    let cleaned = uri.hasPrefix("file://") ? String(uri.dropFirst("file://".count)) : uri
    return UIImage(contentsOfFile: cleaned)
  }

  private func makeBlackImage(side: Int) -> UIImage {
    let size = CGSize(width: side, height: side)
    let renderer = UIGraphicsImageRenderer(size: size)
    return renderer.image { context in
      UIColor.black.setFill()
      context.fill(CGRect(origin: .zero, size: size))
    }
  }
}

File: apps/app/modules/expo-pose-landmarker/android/src/main/java/expo/modules/poselandmarker/ExpoPoseLandmarkerModule.kt
package expo.modules.poselandmarker

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.net.Uri
import android.util.Log
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.core.Delegate
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarker
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File

// MediaPipe Pose Landmarker → COCO-17 mapping. Mirrors
// MEDIAPIPE_TO_COCO17 in landmarkMapping.ts and mediaPipeIndexToCoco17 in
// ExpoPoseLandmarkerModule.swift. Keep all three in lockstep.
private val MEDIAPIPE_TO_COCO17 = mapOf(
  0 to "nose",
  2 to "left_eye",
  5 to "right_eye",
  7 to "left_ear",
  8 to "right_ear",
  11 to "left_shoulder",
  12 to "right_shoulder",
  13 to "left_elbow",
  14 to "right_elbow",
  15 to "left_wrist",
  16 to "right_wrist",
  23 to "left_hip",
  24 to "right_hip",
  25 to "left_knee",
  26 to "right_knee",
  27 to "left_ankle",
  28 to "right_ankle"
)

private const val TAG = "ExpoPoseLandmarker"
private const val MODEL_ASSET_NAME = "pose_landmarker_lite.task"

class ExpoPoseLandmarkerModule : Module() {
  private var landmarker: PoseLandmarker? = null
  private var activeDelegate: Delegate = Delegate.GPU

  override fun definition() = ModuleDefinition {
    Name("ExpoPoseLandmarkerModule")

    OnCreate {
      try {
        initializeLandmarker()
      } catch (error: Throwable) {
        Log.e(TAG, "init failed", error)
      }
    }

    OnDestroy {
      landmarker?.close()
      landmarker = null
    }

    AsyncFunction("detectPoseFromUri") { uri: String, promise: Promise ->
      val current = landmarker
      if (current == null) {
        promise.reject("E_NOT_READY", "PoseLandmarker is not initialized", null)
        return@AsyncFunction
      }

      val bitmap = loadBitmap(uri)
      if (bitmap == null) {
        promise.reject("E_IMAGE_LOAD", "Could not load image at $uri", null)
        return@AsyncFunction
      }

      val mpImage = BitmapImageBuilder(bitmap).build()
      val startedAt = System.nanoTime()

      try {
        val result = current.detect(mpImage)
        val inferenceMs = (System.nanoTime() - startedAt) / 1_000_000.0

        val firstPose = result.landmarks().firstOrNull()
        if (firstPose == null) {
          promise.resolve(null)
          return@AsyncFunction
        }

        val landmarks = mutableListOf<Map<String, Any>>()
        firstPose.forEachIndexed { index, landmark ->
          val name = MEDIAPIPE_TO_COCO17[index] ?: return@forEachIndexed
          // MediaPipe Android returns visibility/presence as Optional<Float>.
          // Prefer presence (model confidence) over visibility (occlusion).
          val confidence = landmark.presence().orElse(landmark.visibility().orElse(0f))
          landmarks.add(
            mapOf(
              "name" to name,
              "x" to landmark.x().toDouble(),
              "y" to landmark.y().toDouble(),
              "confidence" to confidence.toDouble()
            )
          )
        }

        promise.resolve(
          mapOf(
            "landmarks" to landmarks,
            "imageWidth" to bitmap.width.toDouble(),
            "imageHeight" to bitmap.height.toDouble(),
            "inferenceMs" to inferenceMs,
            "delegate" to if (activeDelegate == Delegate.GPU) "gpu" else "cpu"
          )
        )
      } catch (error: Throwable) {
        promise.reject("E_DETECT", error.message ?: "Unknown detection error", error)
      } finally {
        if (!bitmap.isRecycled) bitmap.recycle()
      }
    }

    AsyncFunction("warmUp") { promise: Promise ->
      val current = landmarker
      if (current == null) {
        promise.resolve(null)
        return@AsyncFunction
      }
      val warmUpBitmap = Bitmap.createBitmap(192, 192, Bitmap.Config.ARGB_8888).apply {
        Canvas(this).drawColor(Color.BLACK)
      }
      try {
        current.detect(BitmapImageBuilder(warmUpBitmap).build())
      } catch (error: Throwable) {
        Log.w(TAG, "warmUp inference failed", error)
      } finally {
        if (!warmUpBitmap.isRecycled) warmUpBitmap.recycle()
      }
      promise.resolve(null)
    }

    Function("getActiveDelegate") {
      if (activeDelegate == Delegate.GPU) "gpu" else "cpu"
    }
  }

  // MARK: - Private helpers

  private fun initializeLandmarker() {
    val context = appContext.reactContext
      ?: throw IllegalStateException("React context unavailable")

    fun build(delegate: Delegate): PoseLandmarker {
      val baseOptions = BaseOptions.builder()
        .setModelAssetPath(MODEL_ASSET_NAME)
        .setDelegate(delegate)
        .build()
      val options = PoseLandmarker.PoseLandmarkerOptions.builder()
        .setBaseOptions(baseOptions)
        .setRunningMode(RunningMode.IMAGE)
        .setNumPoses(1)
        .setMinPoseDetectionConfidence(0.5f)
        .setMinPosePresenceConfidence(0.5f)
        .setMinTrackingConfidence(0.5f)
        .build()
      return PoseLandmarker.createFromOptions(context, options)
    }

    landmarker = try {
      activeDelegate = Delegate.GPU
      build(Delegate.GPU)
    } catch (error: Throwable) {
      Log.w(TAG, "GPU delegate failed, falling back to CPU", error)
      activeDelegate = Delegate.CPU
      build(Delegate.CPU)
    }
    Log.i(TAG, "initialized with delegate: $activeDelegate")
  }

  private fun loadBitmap(uri: String): Bitmap? {
    val parsed = Uri.parse(uri)
    val path = if (parsed.scheme == "file") parsed.path else uri
    if (path.isNullOrEmpty()) return null
    val file = File(path)
    if (!file.exists()) return null
    return BitmapFactory.decodeFile(file.absolutePath)
  }
}
